src/hotspot/share/opto/compile.cpp

```diff
@@ -2204,6 +2209,13 @@ void Compile::Optimize() {
       // Try again if candidates exist and made progress
       // by removing some allocations and/or locks.
     } while (progress);
+
+#ifndef PRODUCT
+    Atomic::add(&ConnectionGraph::_no_escape_counter, total_scalar_replaced);
+    Atomic::add(&ConnectionGraph::_no_escape_counter, _local_no_escape_ctr);
```

Comment:
you can merge two atomic operations into one.  Atomic::add(&ConnectionGraph::_no_escape_counter(&ConnectionGraph::_no_escape_counter, _local_no_escape_ctr + total_scalar_replaced); 

This is smart. last revision is the final revision.

Reply:
@navyxliu Thank you so much for your feedback, it is helpful for me :).

As per this issue, since no nodes are ever removed from the list, I think this issue is avoided. If we find a loop back to a node we have already searched and attempt to push this node at line 2620, it will not be added to the list because it already exists.